benchexec/tools/crux.py

```diff
@@ -0,0 +1,86 @@
+# This file is part of BenchExec, a framework for reliable benchmarking:
+# https://github.com/sosy-lab/benchexec
+#
+# SPDX-FileCopyrightText: 2007-2020 Dirk Beyer <https://www.sosy-lab.org>
+#
+# SPDX-License-Identifier: Apache-2.0
+
+import benchexec.tools.template
+import benchexec.result as result
+import re
+from benchexec.tools.sv_benchmarks_util import get_data_model_from_task, ILP32, LP64
+
+
+class Tool(benchexec.tools.template.BaseTool2):
+    """
+    Tool info for Crux (https://crux.galois.com/).
+    """
+
+    def executable(self, tool_locator):
+        return tool_locator.find_executable("crux-llvm-svcomp-driver.sh")
+
+    def name(self):
+        return "Crux"
+
+    def cmdline(self, executable, options, task, rlimits):
+        if task.property_file:
+            options += ["--svcomp-spec", task.property_file]
+        data_model_param = get_data_model_from_task(
+            task, {ILP32: "32bit", LP64: "64bit"}
+        )
+        if data_model_param:
+            options += ["--svcomp-arch", data_model_param]
+        return [executable] + options + list(task.input_files_or_identifier)
```

Comment:
Does your tool actually support both tasks with more than one input file and tasks without an input file? If not, please use `task.input_files` or `task.single_input_file` accordingly, which will provide proper error messages for the user.

Reply:
Crux does support more than one input file, but I don't think it would produce meaningful results if it were not given _any_ input files. (Does that ever happen?) I tried `task.single_input_file`, but that caused strange results like:

```
Minimal command line:
        “['./crux-llvm-svcomp-driver.sh', 'I', 'N', 'P', 'U', 'T', '.', 'F', 'I', 'L', 'E']”
```

When I tried `task.input_files`, it worked as expected:

```
Minimal command line:
        “['./crux-llvm-svcomp-driver.sh', 'INPUT.FILE']”
```

As a result, I ended up choosing `task.input_files`. Let me know if this is correct.

Reply:
> Crux does support more than one input file, but I don't think it would produce meaningful results if it were not given _any_ input files. (Does that ever happen?)

BenchExec supports such tasks, but of course this happens only if the user configures this (not in SV-COMP).

> I tried `task.single_input_file`, but that caused strange results like:
> 
> ```
> Minimal command line:
>         “['./crux-llvm-svcomp-driver.sh', 'I', 'N', 'P', 'U', 'T', '.', 'F', 'I', 'L', 'E']”
> ```

This is just because `single_input_file` is not a list of strings, but a single string (which in Python is a list of characters). One would have to append it to the list of arguments instead of concatenating it. But it doesn't matter, if your tool supports more than one input file, then `task.input_files` is better anyway.